src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs

```diff
@@ -241,5 +249,24 @@ private void InvokeStartCallback()
                 _invokedStartCallback = true;
             }
         }
+
+        public CollectionRulePipelineState GetPipelineState()
+        {
+            CollectionRulePipelineState pipelineStateCopy = new(_stateHolder.ActionCountLimit, _stateHolder.ActionCountSlidingWindowDuration, _stateHolder.RuleDuration);
+
+            lock (_stateHolder)
+            {
+                pipelineStateCopy = new CollectionRulePipelineState(_stateHolder);
+            }
+
+            lock (_stateHolder.ExecutionTimestamps)
+            {
+                pipelineStateCopy.ExecutionTimestamps = new Queue<DateTime>(_stateHolder.ExecutionTimestamps);
```

Comment:
Maybe we can push all this logic into the copy constructor instead of having two under different locks?